hwilib/devices/bitbox02.py

```diff
@@ -0,0 +1,735 @@
+
+        with handle_errors(common_err_msgs["enumerate"], d_data):
+            if not unlocked:
+                raise DeviceNotReadyError(
+                    "Please load wallet to unlock."
+                    if _using_external_gui
+                    else "Please use any subcommand to unlock"
+                )
+            bb02 = client.init()
+            info = bb02.device_info()
+            if not info["initialized"]:
+                raise HWWError("Not initialized", DEVICE_NOT_INITIALIZED)
+            d_data["fingerprint"] = client.get_master_fingerprint_hex()
+
+        result.append(d_data)
+
+        client.close()
+    return result
+
+
+T = TypeVar("T", bound=Callable[..., Any])
+
+
+def bitbox02_exception(f: T) -> T:
+    """
+    Maps bitbox02 library exceptions into a HWI exceptions.
+    """
+
+    @wraps(f)
+    def func(*args, **kwargs):  # type: ignore
+        """ Wraps f, mapping exceptions. """
+        try:
+            return f(*args, **kwargs)
+        except UserAbortException:
+            raise ActionCanceledError("{} canceled".format(f.__name__))
+        except Bitbox02Exception as exc:
+            if exc.code in (ERR_GENERIC, ERR_INVALID_INPUT):
+                raise BadArgumentError("invalid input")
+            raise exc
+
+    return cast(T, func)
+
+
+# This class extends the HardwareWalletClient for BitBox02 specific things
+class Bitbox02Client(HardwareWalletClient):
+    def __init__(self, path: str, password: str = "", expert: bool = False) -> None:
+        """
+        Initializes a new BitBox02 client instance.
+        """
+        super().__init__(path, password=password, expert=expert)
+        if password != "":
+            raise BadArgumentError(
+                "The BitBox02 does not accept a passphrase from the host. Please enable the passphrase option and enter the passphrase on the device during unlock."
+            )
+
+        hid_device = hid.device()
+        hid_device.open_path(path.encode())
+        self.transport = u2fhid.U2FHid(hid_device)
+        self.device_path = path
+
+        # use self.init() to access self.bb02.
+        self.bb02: Optional[bitbox02.BitBox02] = None
+
+        self.noise_config: BitBoxNoiseConfig = CLINoiseConfig()
+
+    def set_noise_config(self, noise_config: BitBoxNoiseConfig) -> None:
+        self.noise_config = noise_config
+
+    def init(self) -> bitbox02.BitBox02:
+        if self.bb02 is not None:
+            return self.bb02
+
+        for device_info in devices.get_any_bitbox02s():
+            if device_info["path"].decode() == self.device_path:
+                bb02 = bitbox02.BitBox02(
+                    transport=self.transport,
+                    device_info=device_info,
+                    noise_config=self.noise_config,
+                )
+                try:
+                    bb02.check_min_version()
+                except FirmwareVersionOutdatedException as exc:
+                    print("WARNING: ", exc)
```

Comment:
Yeah, at some point (also for Electrum) we decided that outdated firmware should be warned only :shrug: i guess the idea was that it might work in most cases anyway, and if it didn't, it would obviously fail and the warning would suffice to indicate the reason :shrug: 

I fixed it to write to stderr and raise the exception.